### PR TITLE
Bolt12/coot/tx submission

### PR DESCRIPTION
**Pull request by bolt12**

### CODE REVIEW

**Review by coot — COMMENTED** [review]
Some remarks on the PR.

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+import Test.QuickCheck
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.QuickCheck (testProperty)
+
+import Control.Concurrent.Class.MonadMVar.Strict qualified as Strict
+import Control.Concurrent.Class.MonadSTM.Strict (StrictTVar)
+import Control.Concurrent.Class.MonadSTM.Strict qualified as Strict
+import Control.Monad (forM)
+import Data.Foldable (traverse_)
+import Data.Void (Void)
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs]: fixed

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+import Ouroboros.Network.DeltaQ (PeerGSV)
+import Ouroboros.Network.TxSubmission.Inbound.Registry
+import Ouroboros.Network.TxSubmission.Inbound.Server (txSubmissionInboundV2)
+import Ouroboros.Network.TxSubmission.Inbound.State
+import Test.Ouroboros.Network.TxSubmission.Common hiding (tests)
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs:70]: Fixed

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+                            ]
+    where
+      shrinkMap1 :: (Ord k, Arbitrary k, Arbitrary v) => Map k v -> [Map k v]
+      shrinkMap1 m
+        | Map.size m <= 1 = [m]
+        | otherwise       = [Map.delete k m | k <- Map.keys m] ++ singletonMaps
+        where
+          singletonMaps = [Map.singleton k v | (k, v) <- Map.toList m]
+
+runTxSubmissionV2
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs:115]: Fixed

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+        -- Run clients and servers
+        withAsyncAll (clients ++ servers) (\asyncs -> k (a, asyncs))
+
+    withAsyncAll :: MonadAsync m => [m a] -> ([Async m a] -> m b) -> m b
+    withAsyncAll xs0 action = go [] xs0
+      where
+        go as []     = action (reverse as)
+        go as (x:xs) = withAsync x (\a -> go (a:as) xs)
+
+txSubmissionV2Simulation :: forall s . TxSubmissionV2State -> IOSim s ([Tx Int], [[Tx Int]])
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs:242]: Fixed

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs [hunk 5]
```diff
@@ -327,6 +328,46 @@ prop_txSubmission st =
...
+                foldr (\tx rr ->
+                        if Map.member tx rr && getTxValid tx
+                           then Map.update (Just . succ @Int) tx rr
+                           else if getTxValid tx
+                                   then Map.insert tx 1 rr
+                                   else rr
+                      )
+                      Map.empty
+                      inmp
+           in resultRepeatedValidTxs === maxRepeatedValidTxs
```
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/TxSubmission/TxSubmissionV2.hs:380]: Yes, that's true, in Diffusion Sim we have the same test but with attenuation.

**Comment by bolt12** [comment]
closed due to it being merged into #4887

**Pull request closed without merging**